modules/core/src/main/java/org/apache/ignite/internal/processors/query/GridQueryProcessor.java

```diff
@@ -340,27 +351,147 @@ public void onCacheReconnect() throws IgniteCheckedException {
+
+        dataMap.put(INLINE_SIZES_DISCO_BAG_KEY, collectSecondaryIndexesInlineSize());
+
+        dataBag.addJoiningNodeData(DiscoveryDataExchangeType.QUERY_PROC.ordinal(), dataMap);
+    }
+
     /** {@inheritDoc} */
     @Override public void onGridDataReceived(DiscoveryDataBag.GridDiscoveryData data) {
-        synchronized (stateMux) {
-            // Preserve proposals.
-            LinkedHashMap<UUID, SchemaProposeDiscoveryMessage> data0 =
-                (LinkedHashMap<UUID, SchemaProposeDiscoveryMessage>)data.commonData();
+        // Preserve proposals.
+        LinkedHashMap<UUID, SchemaProposeDiscoveryMessage> activeProposals =
+            (LinkedHashMap<UUID, SchemaProposeDiscoveryMessage>)data.commonData();
 
-            // Process proposals as if they were received as regular discovery messages.
-            if (data0 != null) {
-                for (SchemaProposeDiscoveryMessage activeProposal : data0.values())
+        // Process proposals as if they were received as regular discovery messages.
+        if (!F.isEmpty(activeProposals)) {
+            synchronized (stateMux) {
+                for (SchemaProposeDiscoveryMessage activeProposal : activeProposals.values())
                     onSchemaProposeDiscovery0(activeProposal);
             }
         }
+
+        if (!F.isEmpty(data.nodeSpecificData())) {
+            Map<String, Integer> indexesInlineSize = secondaryIndexesInlineSize();
+
+            if (!F.isEmpty(indexesInlineSize)) {
+                for (UUID nodeId : data.nodeSpecificData().keySet()) {
+                    Serializable serializable = data.nodeSpecificData().get(nodeId);
+
+                    assert serializable instanceof Map : serializable;
+
+                    Map<String, Serializable> nodeSpecificData = (Map<String, Serializable>)serializable;
+
+                    if (nodeSpecificData.containsKey(INLINE_SIZES_DISCO_BAG_KEY))
+                        checkInlineSizes(indexesInlineSize, (Map<String, Integer>)nodeSpecificData.get(INLINE_SIZES_DISCO_BAG_KEY), nodeId);
+                }
+            }
+        }
+    }
+
+    /**
+     *
+     * @return Information about secondary indexes inline size. Key is a full index name, value is a effective inline size.
+     * @see GridQueryIndexing#secondaryIndexesInlineSize()
+     */
+    public Map<String, Integer> secondaryIndexesInlineSize() {
+        return idx != null ? idx.secondaryIndexesInlineSize() : Collections.emptyMap();
+    }
+
+    /**
+     * Checks that server node with {@code nodeId} is supported {@link IgniteFeatures#CHECK_INDEX_INLINE_SIZES}.
+     *
+     * @param nodeId Node id.
+     * @return {@code true} if node with {@code nodeId} is server node and the node supports feature, {@code false}
+     * otherwise.
+     */
+    private boolean nodeSupportedCheckIndexInlineSize(UUID nodeId) {
+        IgnitePredicate<ClusterNode> nodeFilter = n -> !n.isClient() && !n.isDaemon() && nodeId.equals(n.id());
```

Comment:
"nodeSupportedCheckIndexInlineSize" will return "true" for client nodes, because "allNodesSupport" for empty collection of nodes is "true". Please change javaDoc or fix the method.